### DISCUSSION TITLE
How to calculate VIF (variance inflation factor) with sdmTMB models

### DISCUSSION THREAD

**Discussion by seananderson** [sdmTMB/sdmTMB#202] [Q&A] [answered]
Question by email:

>  I tried to check VIF (variance inflation factor) with sdmTMB models but a few packages that I tried didn't support sdmTMB. Is there a way to check VIF?

**Comment by seananderson** [comment, marked as answer]
Although there's no built-in method currently, it's fairly simple to calculate yourself on the input data frame.

From the [Wikipedia entry](https://en.wikipedia.org/wiki/Variance_inflation_factor):

<img width="189" alt="image" src="https://user-images.githubusercontent.com/19349/234451194-031d06e7-0615-4f1a-b8ff-998a939d88b3.png">

> where $R^2_i$ is the coefficient of determination of the regression equation in step one, with  $X_{i}$ on the left hand side, and all other predictor variables (all the other X variables) on the right hand side.

Here's a little function to do that and an example:

``` r
vif <- function(x) {
  v <- vapply(seq_along(x), function(i) {
    rsq <- summary(lm(x[[i]] ~ . , data = x[,-i, drop = FALSE]))$r.squared
    1 / (1 - rsq)
  }, FUN.VALUE = numeric(1L))
  setNames(v, colnames(x))
}


# subset your data frame to include just the predictors in the model:
d <- mtcars[,c("mpg", "cyl", "disp", "drat")]
vif(d)
#>      mpg      cyl     disp     drat 
#> 4.257023 6.388608 6.349561 2.147019
```

<sup>Created on 2023-04-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ndusit** [comment]
Dear Sean,
I found the 'performance' package that could be calculated VIF.  

> library(performance)
> check_collinearity(M31)
# Check for Multicollinearity

Low Correlation

        Term  VIF       VIF 95% CI Increased SE Tolerance Tolerance 95% CI
  sDist2vill 1.03 [1.00,     1.76]         1.02      0.97     [0.57, 1.00]
    sCorn50m 1.00 [1.00, 1.10e+12]         1.00      1.00     [0.00, 1.00]
 sDist2Water 1.03 [1.00,     1.70]         1.02      0.97     [0.59, 1.00]

Regards,
Dusit
